### DISCUSSION TITLE
Consul service discovery plugin fails when using Erlang/OTP 24

### DISCUSSION THREAD

**Discussion by Enealor** [rabbitmq/rabbitmq-server#3876] [Other] [answered]
Hello! I am trying to use the consul plugin (rabbitmq_peer_discovery_consul) for cluster formation. I followed the steps at https://rabbitmq.com/cluster-formation.html#peer-discovery-consul, but I get the following error when starting the server:
```
2021-12-09 22:16:57.766 [error] <0.192.0> CRASH REPORT Process <0.192.0> with 0 neighbours exited with reason: call to undefined function crypto:hmac(sha512, <<21,135,13,108,158,208,40,165,61,185,76,223,138,4,217,204,82,238,158,155,125,225,178,98,187,13,...>>, [<<240,243,234,37,111,247,141,108,169,76,161,178,141,246,206,162>>,<<0,0,0,1>>], 64) in application_master:init/4 line 142
```
This plugin works when running erlang 1.22, but fails for erlang 1.24. I found that this function (crypto:hmac/4) is deprecated in 1.23 and removed in 1.24.

This currently stops me from using the latest version of OTP.

**Comment by michaelklishin** [comment]
I will convert this issue to a GitHub discussion. Currently GitHub will [automatically close and lock the issue](https://github.com/github/feedback/discussions/3197) even though your question will be transferred and responded to elsewhere. This is to let you know that we do not intend to ignore this but this is how the current GitHub conversion mechanism makes it seem for the users :(

**Comment by michaelklishin** [comment]
Indeed, `crypto:hmac/4` seems to need replacing. I fail to see what in the plugin itself uses it, though, consider posting more than a single line when reporting issues.

**Comment by michaelklishin** [comment]
Another important aspect not mentioned is that the only way to reproduce this is to use an HTTPS endpoint. Which is why neither unit tests nor my local development environment run into this.

Anyhow, I filed https://github.com/rabbitmq/rabbitmq-server/issues/3877 and will add more details as I find them.

**Comment by michaelklishin** [comment]
I cannot find any [direct uses of `crypto:hmac/3` or `crypto:hmac/4` in our codebase](https://github.com/rabbitmq/rabbitmq-server/search?q=crypto%3Ahmac).

The Consul plugin relies on `rabbit_peer_discovery_consul` which uses the Erlang `httpc` client and that's it. SHA-512 listed in the
exception is not the default hashing function anywhere in RabbitMQ, which makes me wonder what exactly the RabbitMQ version and node configuration used are.

Please provide as much information as possible about your environment, configuration and so on. We do not guess in this community.

**Comment by michaelklishin** [comment, marked as answer]
Also keep in mind that [only RabbitMQ 3.8.16 and later versions](https://rabbitmq.com/which-erlang.html) support Erlang 24. Previous versions are not expected to run on it.

**Comment by Enealor** [comment]
Thank you for the information. After reading your last comment, I realized that I did not check versions. I had assumed I was installing latest, and found that I was installing the ubuntu hosted package (version 3.8.2). I updated to 3.9 and the server comes up without the aforementioned error.

Thank you again for your time.
